HIP/hip-584.md

```diff
@@ -0,0 +1,127 @@
+- As a user, I want a library that can be used to re-run historical EVM transactions accurately so that I can inspect 
+and debug the execution of the bytecode logic.
+
+## Specification
+
+### EVM Library Definition
+
+We want the EVM library to be used by multiple clients like the `hedera-services` and `hedera-mirror-node` repos
+and to be able to be extended for a future use from additional clients.
+
+To facilitate differences, we need to define interfaces, which would be inside the library 
+and each client of the library (in this case `hedera-node` and `mirror-node`) would provide its own implementation.
+
+### Mirror Node
+
+Information will be made available via new REST API
+
+#### Read-only queries and gas estimate REST API
+
+A new  `/api/v1/contracts/call` POST REST API will be added to execute read-only queries and gas estimate.
+Query parameter `estimate`(default = false) would determine what the result represents.
+
+The following JSON represents a typical request body:
+```json
+{
+  "block": "latest", 
+  "data": "0x81a73ad500000000000000000000000000000000000000000000000000000000000004e5", 
+  "from": "0x00000000000000000000000000000000000004e2", 
+  "gas": "0x76c0", 
+  "gas_price": "0x76c0", 
+  "to": "0x00000000000000000000000000000000000004e4", 
+  "value": "0x0"
+}
+```
+The following JSON represents a typical response result with the information read in the query:
+```json
+{
+    "gas": "0x3234333230",
+    "result": "0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000037474740000000000000000000000000000000000000000000000000000000000"
+}
+```
+
+## Backwards Compatibility
+
+Since the library would duplicate some code inside `hedera-node`, there would be a need to extract
+and remove this code from `hedera-node` and start using `hedera-evm` as a dependency inside `hedera-node`.
+In this way `hedera-node` and `mirror-node` would reuse the same code base, and we would avoid code duplications.
+
+`Mirror Nodes` that will support the new REST APIs should have enabled importing of `CONTRACT_BYTECODE` and `CONTRACT_STATE_CHANGE` 
+sidecar types. Otherwise, executing the endpoints would result in missing runtime bytecode for execution
+or using stale contract storage data.
+
+## Security Implications
+
+There would be some security implementations for `Mirror Nodes`. Since the invocation of `eth_call` and `eth_gasEstimate` RPC calls 
+would be free of charge and the `Mirror Nodes` don’t have throttle mechanism, some attack vectors would be possible. 
+The Nodes could be overloaded with huge amount of calls or calls invoking smart contract methods with huge gas usage, 
+both of which might take more system usage and slow down the network for other users. A rate limit mechanism will be implemented,
+so that the load put on `Mirror Nodes` to be balanced.
+
+## How to Teach This
+
+Respective documentation will be added.
```

Comment:
Can be more specific with
- Service "EVM Modularization" Design Doc
- "Mirror EVM Archive Node" Design Doc
- Open API doc on Mirror Node
- docs.hedera.com sections for `/api/v1/contracts/call`

Reply:
I believe we can link these corresponding documentations at a later point, so that they would more up-to-date with the latest implementation details for the evm modules.